### PR TITLE
Add a json generator

### PR DESCRIPTION
**Pull request by dmerejkowsky**
Implements #2527

### CODE REVIEW

**Comment by dmerejkowsky** [comment]
Context: We have a C++ library with many dependencies that we build with conan + cmake.

We use `distutils` to compile a Python extension, with code looking like:

```python
   hello = Extension('hello',
                    sources = ['hello.c']
                    extra_objects=hello_libs,
                    include_dirs=[hello_include_paths],
                    language='c++',
    )
```

Since we're using static libraries, we need to find all the `.a` on which `hello.a` depends.

I've tried parsing `conanbuildinfo.txt` "by hand" and then realized it would probably be easier to just write an other generator, hence this pull request.

**Comment by memsharded** [comment]
Thanks for the contribution. Yes, it makes sense, has low risk, and the PR already looks very good. 1.1 is already closed, but it will be in 1.2. If in the meanwhile you want to use it, you can put the generator in a package and consume it from the package instead of built-in: http://docs.conan.io/en/latest/howtos/custom_generators.html

**Comment by dmerejkowsky** [comment]
> If in the meanwhile you want to use it, you can put the generator in a package and consume it from the package instead of built-in:

Good to know, thanks :)

I've also opened a pull request for the changelog and documentation:
https://github.com/conan-io/docs/pull/552

File: conans/client/generators/json_generator.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+class JsonGenerator(Generator):
+    @property
+    def filename(self):
+        return "conaninfo.json"
+
+    @property
+    def content(self):
+        info = {}
+        info["dependencies"] = []
+        for depname, cpp_info in self.deps_build_info.dependencies:
```
Comment (danimtb) [inline conans/client/generators/json_generator.py]: For completitude I would like to see also here the values of `self.deps_env_info`, `self.env_info` and `self._deps_user_info`

File: conans/client/generators/json_generator.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+from conans.model import Generator
+
+
+def serialize_cpp_info(cpp_info):
+    keys = [
+        "version",
+        "description",
+        "rootpath",
+        "sysroot",
+        "include_paths", "lib_paths", "bin_paths", "build_paths", "res_paths",
```
Comment (danimtb) [inline conans/client/generators/json_generator.py]: Better stick to same names as in _conaninfo.txt_: includedirs, libdirs, bindirs...

**Comment by danimtb** [comment]
Hi @dmerejkowsky 
Thank you so much for this PR. It is looking quite good and I really think some users will benefit from this generator. I have added some minor comments in the review to complete the generator.

I think it would be nice to have a functional test using this from `$ conan install name/version@user/channel -g json`, but we could add them too.

Thanks again fot the contribution! 😄

File: conans/client/generators/json_generator.py (hunk 1 above)
Comment (dmerejkowsky) [reply conans/client/generators/json_generator.py]: Done for `deps_env_info` and `deps_user_info`.
I did not manage to figure out when `self.env_info` is not None. Do you have some pointers to help me ?

File: conans/client/generators/json_generator.py (hunk 2 above)
Comment (dmerejkowsky) [reply conans/client/generators/json_generator.py]: Agreed.

**Comment by dmerejkowsky** [comment]
@danimtb Addressed most of your feedback, except for figuring out how to serialize `self.env_info`.


Cheers!

File: conans/client/generators/json_generator.py (hunk 1 above)
Comment (danimtb) [reply conans/client/generators/json_generator.py]: It is environmet information declared in the conanfile that receives the ``$ conan` install`` command. Check for info in the docs: http://docs.conan.io/en/latest/reference/conanfile/methods.html#env-info

Comment (dmerejkowsky) [reply conans/client/generators/json_generator.py]: Hum. I'm missing something then. In the test I wrote I have:

```python
class HelloConan(ConanFile):
    def package_info(self):
        self.env_info.MY_VAR = "foo"
```

But inside the JsonGenerator, `self.conanfile.env_info` is still None.

I do find the 'MY_VAR' key in `self.deps_env_info.vars`, though

Comment (danimtb) [reply conans/client/generators/json_generator.py]: You are right, deps_env_info is where env_info vars are propagated when doing the install: https://github.com/conan-io/conan/blob/207e2c70938b6e74c2279f827d51ec244d701be0/conans/client/installer.py#L453

So I think it is ok the way it is. Sorry for that

Comment (dmerejkowsky) [reply conans/client/generators/json_generator.py]: No problem :)

File: conans/client/generators/json_generator.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    res = {}
+    for key in keys:
+        res[key] = getattr(cpp_info, key)
+    return res
+
+
+class JsonGenerator(Generator):
+    @property
+    def filename(self):
+        return "conaninfo.json"
```
Comment (danimtb) [inline conans/client/generators/json_generator.py]: *conaninfo.json* might be confusing as it is similar to *conaninfo.txt* that gathers the information of a generated pacakge, not info about package dependencies. I suggest using the convention *conanbuildinfo.json*

**Comment by danimtb** [comment]
@dmerejkowsky so one last thin in the review and I think I am done!

**Comment by memsharded** [comment]
Sorry, I have realized that there might be a mistake. Or maybe I understood something wrong. The docs PR in https://github.com/conan-io/docs/pull/552/files shows ``include_paths`` with the absolute path, the same as other generators. But the generator is using the relative ``includedirs``.

It was my mistake, I am proposing a PR to fix it. Thanks!

**Pull request merged**